app/controllers/mixins/breadcrumbs_mixin.rb

```diff
@@ -0,0 +1,144 @@
+          breadcrumbs.push(special_page_breadcrumb(@tagitems || @politems || @ownershipitems || @retireitems))
+          breadcrumbs.push(:title => @right_cell_text) if @sb["action"]
+        end
+      end
+      breadcrumbs.compact
+    end
+
+    def build_breadcrumbs_no_explorer(record_info, record_title)
+      if record_info[record_title]
+        notshow = restful? ? "" : "show"
+        breadcrumb_url = File.join(controller_url, notshow, record_info[:id].to_s)
+        {:url => breadcrumb_url, :title => record_info[record_title]}
+      end
+    end
+
+    # TODO: Replace breadcrumb_url with url to right controller (database parameter to controller)
+    # and add :url => breadcrumb_url parameter to return hash
+    def special_page_breadcrumb(variable)
+      # breadcrumb_url = File.join(controller_url, notshow, variable.first[:id].to_s)
+      # EMS has key instead of name
+      return unless variable
+      if variable.first[:key]
+        {:title => variable.first[:key]}
+      # FloatingIps do not have name
+      elsif floating_ip_address?(variable.first)
+        {:title => variable.first[:address]}
+      else
+        {:title => variable.first[:name]}
+      end
+    end
+
+    # Explorer controller methods
+
+    # Get breadcrumbs from the current tree path
+    def current_tree_path(tree, active_node, path = [])
+      result = []
+      result.replace(path)
+      result.push(:title => tree["text"], :key => tree["key"])
+      return result if tree["key"] == active_node
+      if tree.include?("nodes")
+        tree["nodes"].each do |node|
+          value = current_tree_path(node, active_node, result)
+          return value if value
+        end
+      end
+      nil
+    end
+
+    def build_breadcrumbs_from_tree
+      breadcrumbs = []
+      node = x_tree["active_node"] if x_tree && x_tree["active_node"]
+      trees = build_tree if node
+
+      if trees.present?
+        tree = trees.find { |t| t.instance_variable_get("@name") == x_active_tree }
+        value = current_tree_path(JSON.parse(tree.instance_variable_get("@bs_tree")).first, node) if tree.present?
+        breadcrumbs = value if value
+      end
+
+      breadcrumbs
+    end
+
+    # Helper methods
+
+    def build_tree
+      if features?
+        allowed_features = ApplicationController::Feature.allowed_features(features)
+        return allowed_features.collect { |feature| feature.build_tree(@sb) }
+      end
+      []
+    end
+
+    def accord_name
+      features.find { |f| f.name == x_active_accord }.try(:title)
+    end
+
+    def accord_container
+      features.find { |f| f.name == x_active_accord }.try(:container)
+    end
+
+    # Has controller features
+    def features?
+      defined?(features) && features
+    end
+
+    # User is not on show page
+    def not_show_page?
+      (action_name == "show" && params["display"]) || (action_name != "show")
+    end
+
+    def floating_ip_address?(variable)
```

Comment:
This one is very confusing without a comment where this is needed or a nice condition ;)